File: backend/app/schemas/chapter_workflow.py
# AIMETA P=章节工作流持久状态与HTTP契约|R=checkpoint校验_start_snapshot_command响应|NR=不含正文_prompt_运行时对象|E=ChapterWorkflowStateV1_ChapterWorkflowSnapshot|X=http,internal|A=pydantic_contract|D=pydantic|S=none|RD=./README.ai
"""Serializable state and HTTP contracts for durable Chapter workflows."""

from __future__ import annotations

from typing import Annotated, Any, Literal, Optional
from uuid import UUID

from pydantic import AfterValidator, BaseModel, ConfigDict, Field, model_validator

from .novel import FlowConfig

CHAPTER_WORKFLOW_VERSION_V1 = 1
CHAPTER_WORKFLOW_STATE_SCHEMA_VERSION_V1 = 1
CHAPTER_WORKFLOW_CONTEXT_SCHEMA_VERSION_V1 = 1

ChapterWorkflowNodeKeyV1 = Literal[
    "freeze_context",
    "plan_and_direct",
    "generate_candidates",
    "review_candidates",
    "persist_candidates",
    "waiting_for_selection",
    "finalize_revision",
    "projection_pending",
    "observe_projection",
    "successful",
]
ChapterWorkflowSnapshotNodeKeyV1 = (
    ChapterWorkflowNodeKeyV1
    | Literal[
        "failed",
        "cancelled",
        "superseded",
    ]
)
ChapterWorkflowRunStatus = Literal[
    "queued",
    "running",
    "retry_wait",
    "waiting_for_selection",
    "finalizing",
    "projection_pending",
    "needs_attention",
    "successful",
    "failed",
    "cancelled",
    "superseded",
]
ChapterWorkflowRootJobStatus = Literal[
    "queued",
    "running",
    "retry_wait",
    "waiting",
    "succeeded",
    "failed",
    "dead_letter",
    "needs_attention",
    "cancelled",
]
CHAPTER_WORKFLOW_NODE_KEYS_V1: tuple[ChapterWorkflowNodeKeyV1, ...] = (
    "freeze_context",
    "plan_and_direct",
    "generate_candidates",
    "review_candidates",
    "persist_candidates",
    "waiting_for_selection",
    "finalize_revision",
    "projection_pending",
    "observe_projection",
    "successful",
)

ReferenceKey = Annotated[str, Field(min_length=1, max_length=255)]
ReferenceValue = Annotated[str, Field(min_length=1, max_length=255)]
ChapterVersionId = Annotated[int, Field(ge=1)]
ChapterWorkflowCommandType = Literal[
    "select",
    "retry",
    "retry_external",
    "retry_projection",
    "cancel",
]
ChapterWorkflowCommandStatus = Literal["pending", "applied", "rejected"]


def validate_chapter_workflow_run_id(value: str) -> str:
    """拒绝不能稳定映射为 thread/stream identity 的非规范 UUID。"""

    try:
        parsed = UUID(value)
    except ValueError as error:
        raise ValueError("run_id 必须是规范 UUID") from error
    if str(parsed) != value:
        raise ValueError("run_id 必须是规范 UUID")
    return value


ChapterWorkflowRunId = Annotated[
    str,
    Field(min_length=36, max_length=36),
    AfterValidator(validate_chapter_workflow_run_id),
]


class ChapterWorkflowStateV1(BaseModel):
    """Graph V1 checkpoint；只持久化身份、hash、marker 与结果引用。"""

    model_config = ConfigDict(extra="forbid")

    workflow_version: Literal[1] = 1
    state_schema_version: Literal[1] = 1
    run_id: ChapterWorkflowRunId
    node_key: ChapterWorkflowNodeKeyV1
    context_hash: str = Field(pattern=r"^[0-9a-f]{64}$")
    activity_refs: dict[ReferenceKey, ReferenceValue] = Field(
        default_factory=dict,
        max_length=100,
    )
    result_refs: dict[ReferenceKey, ReferenceValue] = Field(
        default_factory=dict,
        max_length=100,
    )
    candidate_version_ids: list[ChapterVersionId] = Field(
        default_factory=list,
        max_length=100,
    )
    selected_version_id: Optional[int] = Field(default=None, ge=1)
    last_applied_command_id: Optional[str] = Field(
        default=None,
        min_length=36,
        max_length=36,
    )
    target_chapter_revision: Optional[int] = Field(default=None, ge=1)
    error_category: Optional[str] = Field(default=None, min_length=1, max_length=64)

    @classmethod
    def initial(cls, *, run_id: str, context_hash: str) -> "ChapterWorkflowStateV1":
        """从 durable run 身份构造首个 freeze_context checkpoint。"""

        return cls(
            run_id=run_id,
            node_key="freeze_context",
            context_hash=context_hash,
        )


class ChapterWorkflowCommandEnvelope(BaseModel):
    """客户端生成的幂等 command 身份与并发前置条件。"""

    model_config = ConfigDict(extra="forbid")

    command_id: ChapterWorkflowRunId
    type: ChapterWorkflowCommandType
    payload_version: Literal[1] = 1
    payload: dict[str, Any] = Field(default_factory=dict, max_length=16)
    expected_run_revision: int = Field(ge=0)
    expected_chapter_revision: int = Field(ge=0)
    expected_checkpoint_id: str = Field(min_length=1, max_length=512)

    @model_validator(mode="after")
    def validate_command_payload(self) -> "ChapterWorkflowCommandEnvelope":
        payload = self.payload
        if self.type == "select":
            selected_version_id = payload.get("selected_version_id")
            if (
                set(payload) != {"selected_version_id"}
                or isinstance(selected_version_id, bool)
                or not isinstance(selected_version_id, int)
                or selected_version_id < 1
            ):
                raise ValueError("select command payload 无效")
        elif self.type in {"retry", "retry_projection"}:
            if payload:
                raise ValueError(f"{self.type} command payload 必须为空")
        elif self.type == "retry_external":
            if set(payload) != {"activity_key", "acknowledge_possible_duplicate"}:
                raise ValueError("retry_external command payload 无效")
            self._validate_activity_key(payload.get("activity_key"))
            if payload.get("acknowledge_possible_duplicate") is not True:
                raise ValueError("retry_external 必须确认可能重复调用")
        else:
            if set(payload) not in (set(), {"activity_key"}):
                raise ValueError("cancel command payload 无效")
            if "activity_key" in payload:
                self._validate_activity_key(payload.get("activity_key"))
        return self

    @staticmethod
    def _validate_activity_key(value: object) -> None:
        if not isinstance(value, str) or not value.strip() or len(value) > 128:
            raise ValueError("activity_key 必须为 1 到 128 个字符")


class ChapterWorkflowSnapshot(BaseModel):
    """供 command/API 复用的当前 run、Chapter 与事件游标快照。"""

    model_config = ConfigDict(extra="forbid")

    run_id: ChapterWorkflowRunId
    root_job_id: ChapterWorkflowRunId
    project_id: str = Field(min_length=1, max_length=36)
    chapter_id: int = Field(ge=1)
    chapter_number: int = Field(ge=1)
    base_revision: int = Field(ge=0)
    current_chapter_revision: int = Field(ge=0)
    workflow_version: Literal[1]
    state_schema_version: Literal[1]
    context_schema_version: Literal[1]
    status: ChapterWorkflowRunStatus
    root_job_status: ChapterWorkflowRootJobStatus
    node_key: ChapterWorkflowSnapshotNodeKeyV1
    checkpoint_id: Optional[str] = Field(default=None, max_length=512)
    progress: int = Field(ge=0, le=100)
    row_revision: int = Field(ge=0)
    is_active: bool
    successor_run_id: Optional[ChapterWorkflowRunId] = None
    error_category: Optional[str] = Field(default=None, max_length=64)
    public_error: Optional[str] = Field(default=None, max_length=512)
    allowed_commands: list[ChapterWorkflowCommandType]
    retry_activity_key: Optional[str] = Field(max_length=128)
    resume_cursor: int = Field(ge=0)


class ChapterWorkflowStartRequest(BaseModel):
    """创建或复用当前章节 active workflow 的公开输入。"""

    model_config = ConfigDict(extra="forbid")

    project_id: str = Field(min_length=1, max_length=36)
    chapter_number: int = Field(ge=1)
    writing_notes: Optional[str] = None
    flow_config: FlowConfig = Field(default_factory=FlowConfig)


class ChapterWorkflowConnection(BaseModel):
    """可恢复的公开 workflow 快照与 opaque durable event stream 地址。"""

    model_config = ConfigDict(extra="forbid")

    snapshot: ChapterWorkflowSnapshot
    events_url: str = Field(min_length=1, max_length=2_048)


class ChapterWorkflowStartResponse(ChapterWorkflowConnection):
    """start 结果在可恢复连接之外标识本次是否创建 run。"""

    created: bool


class ChapterWorkflowCommandResponse(BaseModel):
    """已持久化 command 的公开状态与提交后的当前快照。"""

    model_config = ConfigDict(extra="forbid")

    command_id: ChapterWorkflowRunId
    type: ChapterWorkflowCommandType
    status: ChapterWorkflowCommandStatus
    snapshot: ChapterWorkflowSnapshot


class ChapterWorkflowCommandConflictDetail(BaseModel):
    """409 冲突携带稳定原因和同事务事实之后读取的当前快照。"""

    model_config = ConfigDict(extra="forbid")

    reason_code: str = Field(min_length=1, max_length=64)
    current_snapshot: ChapterWorkflowSnapshot


class ChapterWorkflowCommandConflictResponse(BaseModel):
    """FastAPI 默认 error envelope 中的 typed 409 detail。"""

    model_config = ConfigDict(extra="forbid")

    detail: ChapterWorkflowCommandConflictDetail

File: frontend/src/api/chapterWorkflow.ts
// AIMETA P=章节工作流API_快照契约解码|R=current_start_command请求与运行时校验|NR=不持有服务端缓存或UI状态|E=api:chapter-workflow|X=internal|A=ChapterWorkflowAPI|D=fetch,openapi|S=net|RD=./README.ai
import { API_BASE_URL, API_PREFIX } from './base'
import { authJson } from './client'
import type { components } from './generated/schema'
import type { HttpRequestOptions } from './http'

export type ChapterWorkflowCommandConflictDetail =
  components['schemas']['ChapterWorkflowCommandConflictDetail']
export type ChapterWorkflowCommandConflictResponse =
  components['schemas']['ChapterWorkflowCommandConflictResponse']
export type ChapterWorkflowCommandEnvelope =
  components['schemas']['ChapterWorkflowCommandEnvelope']
export type ChapterWorkflowCommandResponse =
  components['schemas']['ChapterWorkflowCommandResponse']
export type ChapterWorkflowConnection = components['schemas']['ChapterWorkflowConnection']
export type ChapterWorkflowSnapshot = components['schemas']['ChapterWorkflowSnapshot']
export type ChapterWorkflowStartRequest = components['schemas']['ChapterWorkflowStartRequest']
export type ChapterWorkflowStartResponse = components['schemas']['ChapterWorkflowStartResponse']

export type ChapterWorkflowCommand = ChapterWorkflowCommandEnvelope['type']
export type ChapterWorkflowCommandStatus = ChapterWorkflowCommandResponse['status']
export type ChapterWorkflowNodeKey = ChapterWorkflowSnapshot['node_key']
export type ChapterWorkflowRootJobStatus = ChapterWorkflowSnapshot['root_job_status']
export type ChapterWorkflowStatus = ChapterWorkflowSnapshot['status']

export interface ChapterWorkflowScope {
  projectId: string
  chapterNumber: number
}

interface ChapterWorkflowSnapshotExpectation {
  scope?: ChapterWorkflowScope
  runId?: string
}

interface ChapterWorkflowCommandExpectation extends ChapterWorkflowSnapshotExpectation {
  commandId?: string
  commandType?: ChapterWorkflowCommand
}

export type ChapterWorkflowContractErrorCode =
  | 'identity_mismatch'
  | 'malformed'
  | 'scope_mismatch'
  | 'unsupported_version'

export class ChapterWorkflowContractError extends Error {
  readonly code: ChapterWorkflowContractErrorCode
  readonly reason: string

  constructor(code: ChapterWorkflowContractErrorCode, reason: string) {
    const message = code === 'unsupported_version'
      ? '章节工作流数据版本不受支持'
      : code === 'scope_mismatch' || code === 'identity_mismatch'
        ? '章节工作流数据身份不匹配'
        : '章节工作流数据格式无效'
    super(message)
    this.name = 'ChapterWorkflowContractError'
    this.code = code
    this.reason = reason
  }
}

const exactStringValues =
  <Expected extends string>() =>
  <const Values extends readonly Expected[]>(
    values: Values & ([Expected] extends [Values[number]] ? unknown : never),
  ) => values

export const CHAPTER_WORKFLOW_STATUS_VALUES = exactStringValues<ChapterWorkflowStatus>()([
  'queued',
  'running',
  'retry_wait',
  'waiting_for_selection',
  'finalizing',
  'projection_pending',
  'needs_attention',
  'successful',
  'failed',
  'cancelled',
  'superseded',
] as const)

export const CHAPTER_WORKFLOW_ROOT_JOB_STATUS_VALUES =
  exactStringValues<ChapterWorkflowRootJobStatus>()([
    'queued',
    'running',
    'retry_wait',
    'waiting',
    'succeeded',
    'failed',
    'dead_letter',
    'needs_attention',
    'cancelled',
  ] as const)

export const CHAPTER_WORKFLOW_NODE_KEY_VALUES = exactStringValues<ChapterWorkflowNodeKey>()([
  'freeze_base_context',
  'retrieve_context',
  'plan_chapter',
  'generate_candidate_1',
  'generate_candidate_2',
  'review_candidates',
  'refine_candidate',
  'enhance_content',
  'repair_consistency',
  'optimize_style',
  'enrich_content',
  'compress_candidate',
  'persist_drafts',
  'wait_for_selection',
  'finalize_revision',
  'wait_for_projections',
  'reconcile_projections',
  'successful',
  'failed',
  'cancelled',
  'superseded',
] as const)

export const CHAPTER_WORKFLOW_COMMAND_VALUES = exactStringValues<ChapterWorkflowCommand>()([
  'select',
  'retry',
  'retry_external',
  'retry_projection',
  'cancel',
] as const)

export const CHAPTER_WORKFLOW_COMMAND_STATUS_VALUES =
  exactStringValues<ChapterWorkflowCommandStatus>()([
    'pending',
    'applied',
    'rejected',
  ] as const)

const WORKFLOW_BASE = `${API_BASE_URL}${API_PREFIX}/writer/chapter-workflows`
const CANONICAL_UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/

const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null && !Array.isArray(value)

const isIntegerInRange = (value: unknown, min: number, max = Number.MAX_SAFE_INTEGER) =>
  typeof value === 'number' && Number.isInteger(value) && value >= min && value <= max

const isStringInRange = (value: unknown, min: number, max: number) =>
  typeof value === 'string' && value.length >= min && value.length <= max

const isOptionalNullableString = (
  value: unknown,
  max: number,
  min = 0,
) => value === undefined || value === null || isStringInRange(value, min, max)

const isCanonicalUuid = (value: unknown): value is string =>
  typeof value === 'string' && CANONICAL_UUID_PATTERN.test(value)

const isOptionalNullableCanonicalUuid = (value: unknown) =>
  value === undefined || value === null || isCanonicalUuid(value)

const isStringValue = <Value extends string>(
  value: unknown,
  values: readonly Value[],
): value is Value => typeof value === 'string' && (values as readonly string[]).includes(value)

const failContract = (
  code: ChapterWorkflowContractErrorCode,
  reason: string,
): never => {
  throw new ChapterWorkflowContractError(code, reason)
}

const requireRecord = (payload: unknown, reason: string): Record<string, unknown> => {
  if (!isRecord(payload)) {
    throw new ChapterWorkflowContractError('malformed', reason)
  }
  return payload
}

const requireSupportedVersions = (value: Record<string, unknown>) => {
  const versionFields = [
    'workflow_version',
    'state_schema_version',
    'context_schema_version',
  ] as const
  for (const field of versionFields) {
    if (!isIntegerInRange(value[field], 0)) {
      failContract('malformed', field)
    }
    if (value[field] !== 1) {
      failContract('unsupported_version', field)
    }
  }
}

export const decodeChapterWorkflowSnapshot = (
  payload: unknown,
  expectation: ChapterWorkflowSnapshotExpectation = {},
): ChapterWorkflowSnapshot => {
  const value = requireRecord(payload, 'snapshot')
  requireSupportedVersions(value)
  if (
    !isCanonicalUuid(value.run_id)
    || !isCanonicalUuid(value.root_job_id)
    || !isStringInRange(value.project_id, 1, 36)
    || !isIntegerInRange(value.chapter_id, 1)
    || !isIntegerInRange(value.chapter_number, 1)
    || !isIntegerInRange(value.base_revision, 0)
    || !isIntegerInRange(value.current_chapter_revision, 0)
    || !isStringValue(value.status, CHAPTER_WORKFLOW_STATUS_VALUES)
    || !isStringValue(value.root_job_status, CHAPTER_WORKFLOW_ROOT_JOB_STATUS_VALUES)
    || !isStringValue(value.node_key, CHAPTER_WORKFLOW_NODE_KEY_VALUES)
    || !isOptionalNullableString(value.checkpoint_id, 512)
    || !isIntegerInRange(value.progress, 0, 100)
    || !isIntegerInRange(value.row_revision, 0)
    || typeof value.is_active !== 'boolean'
    || !isOptionalNullableCanonicalUuid(value.successor_run_id)
    || !isOptionalNullableString(value.error_category, 64)
    || !isOptionalNullableString(value.public_error, 512)
    || !Array.isArray(value.allowed_commands)
    || !value.allowed_commands.every((command) =>
      isStringValue(command, CHAPTER_WORKFLOW_COMMAND_VALUES))
    || !Object.prototype.hasOwnProperty.call(value, 'retry_activity_key')
    || !(value.retry_activity_key === null
      || isStringInRange(value.retry_activity_key, 1, 128))
    || !isIntegerInRange(value.resume_cursor, 0)
  ) {
    failContract('malformed', 'snapshot')
  }
  if (
    expectation.scope
    && (
      value.project_id !== expectation.scope.projectId
      || value.chapter_number !== expectation.scope.chapterNumber
    )
  ) {
    failContract('scope_mismatch', 'project_or_chapter')
  }
  if (expectation.runId && value.run_id !== expectation.runId) {
    failContract('identity_mismatch', 'run_id')
  }
  return value as ChapterWorkflowSnapshot
}

export const decodeChapterWorkflowConnection = (
  payload: unknown,
  scope: ChapterWorkflowScope,
): ChapterWorkflowConnection => {
  const value = requireRecord(payload, 'connection')
  if (!isStringInRange(value.events_url, 1, 2_048)) {
    failContract('malformed', 'connection')
  }
  return {
    events_url: value.events_url as string,
    snapshot: decodeChapterWorkflowSnapshot(value.snapshot, { scope }),
  }
}

export const decodeCurrentChapterWorkflow = (
  payload: unknown,
  scope: ChapterWorkflowScope,
): ChapterWorkflowConnection | null =>
  payload === null ? null : decodeChapterWorkflowConnection(payload, scope)

export const decodeChapterWorkflowStartResponse = (
  payload: unknown,
  scope: ChapterWorkflowScope,
): ChapterWorkflowStartResponse => {
  const value = requireRecord(payload, 'start_response')
  if (typeof value.created !== 'boolean') {
    failContract('malformed', 'start_response')
  }
  return {
    ...decodeChapterWorkflowConnection(value, scope),
    created: value.created as boolean,
  }
}

export const decodeChapterWorkflowCommandResponse = (
  payload: unknown,
  expectation: ChapterWorkflowCommandExpectation,
): ChapterWorkflowCommandResponse => {
  const value = requireRecord(payload, 'command_response')
  if (
    !isCanonicalUuid(value.command_id)
    || !isStringValue(value.type, CHAPTER_WORKFLOW_COMMAND_VALUES)
    || !isStringValue(value.status, CHAPTER_WORKFLOW_COMMAND_STATUS_VALUES)
  ) {
    failContract('malformed', 'command_response')
  }
  if (expectation.commandId && value.command_id !== expectation.commandId) {
    failContract('identity_mismatch', 'command_id')
  }
  if (expectation.commandType && value.type !== expectation.commandType) {
    failContract('identity_mismatch', 'command_type')
  }
  return {
    command_id: value.command_id as string,
    type: value.type as ChapterWorkflowCommand,
    status: value.status as ChapterWorkflowCommandStatus,
    snapshot: decodeChapterWorkflowSnapshot(value.snapshot, expectation),
  }
}

export const decodeChapterWorkflowCommandConflict = (
  payload: unknown,
  expectation: ChapterWorkflowSnapshotExpectation,
): ChapterWorkflowCommandConflictResponse => {
  const value = requireRecord(payload, 'command_conflict')
  const detail = requireRecord(value.detail, 'command_conflict')
  if (!isStringInRange(detail.reason_code, 1, 64)) {
    failContract('malformed', 'command_conflict_reason')
  }
  return {
    detail: {
      reason_code: detail.reason_code as string,
      current_snapshot: decodeChapterWorkflowSnapshot(
        detail.current_snapshot,
        expectation,
      ),
    },
  }
}

const request = (url: string, options: HttpRequestOptions = {}) =>
  authJson<unknown>(url, {
    ...options,
    fallbackErrorMessage: '章节工作流接口请求失败',
  })

export class ChapterWorkflowAPI {
  static async getCurrent(
    scope: ChapterWorkflowScope,
    options: HttpRequestOptions = {},
  ): Promise<ChapterWorkflowConnection | null> {
    const params = new URLSearchParams({
      project_id: scope.projectId,
      chapter_number: String(scope.chapterNumber),
    })
    const payload = await request(`${WORKFLOW_BASE}/current?${params.toString()}`, options)
    return decodeCurrentChapterWorkflow(payload, scope)
  }

  static async getSnapshot(
    scope: ChapterWorkflowScope,
    runId: string,
    options: HttpRequestOptions = {},
  ): Promise<ChapterWorkflowSnapshot> {
    const payload = await request(`${WORKFLOW_BASE}/${encodeURIComponent(runId)}`, options)
    return decodeChapterWorkflowSnapshot(payload, { scope, runId })
  }

  static async start(
    input: ChapterWorkflowStartRequest,
    options: HttpRequestOptions = {},
  ): Promise<ChapterWorkflowStartResponse> {
    const payload = await request(WORKFLOW_BASE, {
      ...options,
      method: 'POST',
      body: JSON.stringify(input),
    })
    return decodeChapterWorkflowStartResponse(payload, {
      projectId: input.project_id,
      chapterNumber: input.chapter_number,
    })
  }

  static async submitCommand(
    scope: ChapterWorkflowScope,
    runId: string,
    command: ChapterWorkflowCommandEnvelope,
    options: HttpRequestOptions = {},
  ): Promise<ChapterWorkflowCommandResponse> {
    const payload = await request(
      `${WORKFLOW_BASE}/${encodeURIComponent(runId)}/commands`,
      {
        ...options,
        method: 'POST',
        body: JSON.stringify(command),
      },
    )
    return decodeChapterWorkflowCommandResponse(payload, {
      scope,
      runId,
      commandId: command.command_id,
      commandType: command.type,
    })
  }
}
